bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossHRV.java

```diff
@@ -0,0 +1,323 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.danfossairunit.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+import java.time.DateTimeException;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.types.*;
+import org.eclipse.smarthome.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link DanfossHRV} class does the actual network communication with the air unit.
+ *
+ * @author Ralf Duckstein - Initial contribution
+ * @author Robert Bach - heavy refactorings
+ */
+
+@NonNullByDefault
+public class DanfossHRV implements Commands {
+
+    private final Logger logger = LoggerFactory.getLogger(DanfossHRV.class);
+
+    private InetAddress inetAddr;
+    private int port;
+    private @Nullable Socket socket;
+    private @Nullable OutputStream oStream;
+    private @Nullable InputStream iStream;
+
+    public DanfossHRV(InetAddress inetAddr, int port) {
+        this.inetAddr = inetAddr;
+        this.port = port;
+    }
+
+    public synchronized void connect() throws IOException {
```

Comment:
🤦‍♂ This isn't what I meant by synchronizing connect()...disconnect() logic. I will elaborate on what I meant.
Right now all of your communication logic happens in short bursts: connect->send/receive messages -> disconnect. 
The issue I'm trying to point out is what happens when, lets say, a RefreshCommand and a your scheduled updateAllChannels() happen at the same time? Whichever one calls `connect()` first will have its `socket` instance overwritten by the second call to `connect()`, this will cause a number of problems depending on when the second call occurs. The most important problem is that if you overwrite the socket instance before closing it, it will never get closed! In the worst case this could eventually lead to crashing either openhab or the airunit you are talking to.

Reply:
Ah ok, got your point. Very good one indeed. I refactored the whole connection & request handling to encapsulate it in one class (which is used by the the scheduled job and the updateChannel method in turn). The connect & disconnect & request methods are synchronized. 
A broken connection will block one of the scheduler threads until the timeout, though. I take it the scheduler threads are shared between all bindings? My assumption for now is that this won't pose a problem. If it does, is there a "proper" way to handle that?